### ISSUE TITLE
Writers get blocked in flush() call (system test failure)

### ISSUE THREAD

**Issue by RaulGracia** [pravega/pravega#3313]
**Problem description**
Executing system tests in `r0.4` (Docker Swarm), I observed the following behavior in `readWriteAndAutoScaleWithFailoverTest`:
- The test fails due to a timeout, as readers get stuck waiting for the termination condition to be met.
- One of the conditions for a reader to finish its execution is that `testState.getEventReadCount() == testState.getEventWrittenCount()`. However, what we find at the end of the test is that _there is one more read events than written events_:
```
2019-01-24 18:23:09,453 706444 [ReadWriteAndAutoScaleWithFailoverTest-main-1] DEBUG i.p.t.system.AbstractReadWriteTest - Event write count 16736
...
2019-01-24 18:23:11,046 708037 [ReadWriteAndAutoScaleWithFailoverTest-main-5] DEBUG i.p.t.system.AbstractReadWriteTest - Event read count 16737
```
- The cause for this problem is on the writer side. That is, the writer first writes an event, then flushes the event and then increments the global counter of written events:
```
(AbstractReadWriteTest.java)
final String eventContent = uniqueRoutingKey + RK_VALUE_SEPARATOR + seqNumber;
log.debug("Event write count before write call {}", testState.getEventWrittenCount());
writer.writeEvent(uniqueRoutingKey, eventContent);
log.debug("Event write count before flush {}", testState.getEventWrittenCount());
writer.flush();
testState.incrementTotalWrittenEvents();
log.debug("Writing event {}", eventContent);
log.debug("Event write count {}", testState.getEventWrittenCount());
seqNumber++;
```
If we count the number of messages in the log, we see the following:
`Event write count before write call `: 16737 occurrences
`Event write count before flush`: 16737 occurrences
`Writing event `:  16736 occurrences
 
- Therefore, in this particular case, 1 writer got stuck in the `flush()` method indefinitely. Actually, if we look for the event that is not logged as written (`a40227b8-7aff-4bd1-a0f9-f522d3445d4a:445`), we see that it was created when the number of written events was (approx.) `15809`. That is, a while before the test finishes. 

- Apparently, despite the writer gets blocked flushing the in-flight events and cannot increment the global write counter, the event gets actually written as the reader is able to read it. This causes the mismatch of read/written events in the test.

At first glance, it looks like the behavior we are witnessing is not what the application may expect: that is, the in-flight events were written to the stream but the flush did not unblocked the caller.

**Problem location**
`EventStreamWriterImpl.java`

**Suggestions for an improvement**
Initially, it seems that there is a problem in the blocking behavior of the `flush()` call that needs to be fixed.

**Comment by RaulGracia** [comment]
To elaborate a bit more on this issue, let me describe what happens no the client side:
1) First, writer `9a131fe4-2fd2-48f4-b69a-38f3aac98cb2` writes event `a40227b8-7aff-4bd1-a0f9-f522d3445d4a:444` successfully (the previous to the one that blocked the writer):
```
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.t.system.AbstractReadWriteTest - Event write count before write call 15808
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] TRACE i.p.c.s.i.SegmentOutputStreamImpl - Adding event 1442 to inflight on writer 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] TRACE i.p.c.s.i.SegmentOutputStreamImpl - Sending append request: Append(segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, eventNumber=1442, eventCount=1, data=CompositeByteBuf(ridx: 0, widx: 55, cap: 55, components=2), expectedLength=null)
2019-01-24 18:22:29,426 666417 [epollEventLoopGroup-5-20] TRACE i.p.s.protocol.netty.CommandEncoder - Encoding message to send over the wire Append(segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, eventNumber=1442, eventCount=1, data=CompositeByteBuf(ridx: 0, widx: 55, cap: 55, components=2), expectedLength=null)
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.t.system.AbstractReadWriteTest - Event write count before flush 15808
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 343d22e3-3bce-43d0-9afc-a334fe38542f with 0 inflight events
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 68af0c48-f8be-4171-9d2a-18b106d1f221 with 0 inflight events
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 9200ea5b-0c39-4b80-b5b2-8f9f29ee4ad2 with 0 inflight events
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 88ae14a0-d65a-40d7-bdb6-feff93ac92d3 with 0 inflight events
2019-01-24 18:22:29,426 666417 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2 with 1 inflight events
2019-01-24 18:22:29,426 666417 [epollEventLoopGroup-5-20] TRACE i.p.s.protocol.netty.CommandEncoder - Encoding message to send over the wire WireCommands.KeepAlive(type=KEEP_ALIVE)
2019-01-24 18:22:29,431 666422 [epollEventLoopGroup-5-20] TRACE i.p.s.protocol.netty.CommandDecoder - Decode a message on connection: 10.0.0.17/10.0.0.17:12345. Message was WireCommands.DataAppended(type=DATA_APPENDED, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, eventNumber=1442, previousEventNumber=1441)
2019-01-24 18:22:29,431 666422 [epollEventLoopGroup-5-20] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 10.0.0.17 processing reply: WireCommands.DataAppended(type=DATA_APPENDED, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, eventNumber=1442, previousEventNumber=1441)
2019-01-24 18:22:29,431 666422 [epollEventLoopGroup-5-20] TRACE i.p.c.s.i.SegmentOutputStreamImpl - Received ack: WireCommands.DataAppended(type=DATA_APPENDED, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, eventNumber=1442, previousEventNumber=1441)
2019-01-24 18:22:29,431 666422 [epollEventLoopGroup-5-20] TRACE i.p.c.s.i.SegmentOutputStreamImpl - Inflight empty for writer 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2
2019-01-24 18:22:29,431 666422 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 4e0537f6-278b-444a-b628-71aba74d3ccc with 0 inflight events
2019-01-24 18:22:29,431 666422 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.t.system.AbstractReadWriteTest - Writing event a40227b8-7aff-4bd1-a0f9-f522d3445d4a:444
2019-01-24 18:22:29,431 666422 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.t.system.AbstractReadWriteTest - Event write count 15809
```

2. Then, the writer tries to write the failed event (`a40227b8-7aff-4bd1-a0f9-f522d3445d4a:445`), but gets a `WRONG_HOST` error (probably due to the failovers induced by the test) when connecting to segment `testReadWriteAndAutoScaleStream/7.#epoch.3`:
```
2019-01-24 18:22:29,531 666522 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.t.system.AbstractReadWriteTest - Event write count before flush 15812
2019-01-24 18:22:29,531 666522 [ReadWriteAndAutoScaleWithFailoverTest-main-3] TRACE i.p.c.s.i.SegmentOutputStreamImpl - Adding event 1443 to inflight on writer 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2
2019-01-24 18:22:29,531 666522 [ReadWriteAndAutoScaleWithFailoverTest-main-3] TRACE i.p.c.s.i.SegmentOutputStreamImpl - Sending append request: Append(segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, eventNumber=1443, eventCount=1, data=CompositeByteBuf(ridx: 0, widx: 55, cap: 55, components=2), expectedLength=null)
2019-01-24 18:22:29,531 666522 [epollEventLoopGroup-5-20] TRACE i.p.s.protocol.netty.CommandEncoder - Encoding message to send over the wire Append(segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, eventNumber=1443, eventCount=1, data=CompositeByteBuf(ridx: 0, widx: 55, cap: 55, components=2), expectedLength=null)
2019-01-24 18:22:29,531 666522 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.t.system.AbstractReadWriteTest - Event write count before flush 15812
2019-01-24 18:22:29,531 666522 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 343d22e3-3bce-43d0-9afc-a334fe38542f with 0 inflight events
2019-01-24 18:22:29,532 666523 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 68af0c48-f8be-4171-9d2a-18b106d1f221 with 0 inflight events
2019-01-24 18:22:29,532 666523 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 9200ea5b-0c39-4b80-b5b2-8f9f29ee4ad2 with 0 inflight events
2019-01-24 18:22:29,532 666523 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 88ae14a0-d65a-40d7-bdb6-feff93ac92d3 with 0 inflight events
2019-01-24 18:22:29,532 666523 [ReadWriteAndAutoScaleWithFailoverTest-main-3] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2 with 1 inflight events
2019-01-24 18:22:29,532 666523 [epollEventLoopGroup-5-20] TRACE i.p.s.protocol.netty.CommandEncoder - Encoding message to send over the wire WireCommands.KeepAlive(type=KEEP_ALIVE)
2019-01-24 18:22:29,532 666523 [epollEventLoopGroup-5-20] TRACE i.p.s.protocol.netty.CommandDecoder - Decode a message on connection: 10.0.0.17/10.0.0.17:12345. Message was WireCommands.WrongHost(type=WRONG_HOST, requestId=1443, segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, correctHost=, serverStackTrace=)
2019-01-24 18:22:29,532 666523 [epollEventLoopGroup-5-20] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 10.0.0.17 processing reply: WireCommands.WrongHost(type=WRONG_HOST, requestId=1443, segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, correctHost=, serverStackTrace=)
2019-01-24 18:22:29,532 666523 [epollEventLoopGroup-5-20] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Failing connection for writer 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2 with exception io.pravega.shared.protocol.netty.ConnectionFailedException: WireCommands.WrongHost(type=WRONG_HOST, requestId=1443, segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, correctHost=, serverStackTrace=)
2019-01-24 18:22:29,532 666523 [epollEventLoopGroup-5-20] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Handling exception io.pravega.shared.protocol.netty.ConnectionFailedException: WireCommands.WrongHost(type=WRONG_HOST, requestId=1443, segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, correctHost=, serverStackTrace=) for connection io.pravega.client.netty.impl.ClientConnectionInboundHandler@5271761e on writer 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2. SetupCompleted: true, Closed: false
```

3. The client retries to connect to the segment for 20 seconds (approx.):
```
2019-01-24 18:22:29,541 666532 [epollEventLoopGroup-5-57] WARN  i.p.c.s.i.SegmentOutputStreamImpl - Connection for segment testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3 on writer 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2 failed due to: class io.pravega.shared.protocol.netty.ConnectionFailedException
2019-01-24 18:22:29,541 666532 [epollEventLoopGroup-5-57] WARN  i.p.c.s.i.SegmentOutputStreamImpl - 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2 Failed to connect: 
java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException
...
2019-01-24 18:22:51,018 688009 [epollEventLoopGroup-5-14] WARN  i.p.c.s.i.SegmentOutputStreamImpl - Connection for segment testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3 on writer 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2 failed due to: class io.pravega.shared.protocol.netty.ConnectionFailedException
2019-01-24 18:22:51,018 688009 [epollEventLoopGroup-5-14] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - (Re)connect invoked, Segment: testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerID: 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2
```

4. Until it connects to the segment and writes the event:
```
2019-01-24 18:22:56,018 693009 [clientInternal-26] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Running reconnect for segment:testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3 writerID: 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2
2019-01-24 18:22:56,018 693009 [clientInternal-26] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Fetching endpoint for segment testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerID: 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2
...
2019-01-24 18:22:56,021 693012 [clientInternal-33] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Establishing connection to PravegaNodeUri(endpoint=10.0.0.15, port=12345) for testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerID: 9a131fe4-2fd2-48f4-b69a-38f3aac98cb2
2019-01-24 18:22:56,021 693012 [epollEventLoopGroup-5-27] INFO  i.p.c.n.i.ClientConnectionInboundHandler - Connection established ChannelHandlerContext(ClientConnectionInboundHandler#0, [id: 0x80631448]) 
2019-01-24 18:22:56,021 693012 [epollEventLoopGroup-5-27] TRACE i.p.s.protocol.netty.CommandEncoder - Encoding message to send over the wire WireCommands.Hello(type=HELLO, highVersion=6, lowVersion=5)
2019-01-24 18:22:56,022 693013 [epollEventLoopGroup-5-27] DEBUG i.p.c.n.impl.ConnectionFactoryImpl - Connect operation completed for channel:80631448, local address:/10.0.0.18:35794, remote address:10.0.0.15/10.0.0.15:12345
2019-01-24 18:22:56,022 693013 [epollEventLoopGroup-5-27] TRACE i.p.s.protocol.netty.CommandEncoder - Encoding message to send over the wire WireCommands.SetupAppend(type=SETUP_APPEND, requestId=5, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, delegationToken=)
2019-01-24 18:22:56,022 693013 [epollEventLoopGroup-5-27] TRACE i.p.s.protocol.netty.CommandDecoder - Decode a message on connection: 10.0.0.15/10.0.0.15:12345. Message was WireCommands.Hello(type=HELLO, highVersion=6, lowVersion=5)
2019-01-24 18:22:56,022 693013 [epollEventLoopGroup-5-27] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 10.0.0.15 processing reply: WireCommands.Hello(type=HELLO, highVersion=6, lowVersion=5)
2019-01-24 18:22:56,022 693013 [epollEventLoopGroup-5-27] INFO  i.p.s.p.netty.FailingReplyProcessor - Received hello: WireCommands.Hello(type=HELLO, highVersion=6, lowVersion=5)
2019-01-24 18:22:56,023 693014 [epollEventLoopGroup-5-27] TRACE i.p.s.protocol.netty.CommandDecoder - Decode a message on connection: 10.0.0.15/10.0.0.15:12345. Message was WireCommands.AppendSetup(type=APPEND_SETUP, requestId=5, segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, lastEventNumber=1442)
2019-01-24 18:22:56,023 693014 [epollEventLoopGroup-5-27] DEBUG i.p.c.n.i.ClientConnectionInboundHandler - 10.0.0.15 processing reply: WireCommands.AppendSetup(type=APPEND_SETUP, requestId=5, segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, lastEventNumber=1442)
2019-01-24 18:22:56,023 693014 [epollEventLoopGroup-5-27] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Received AppendSetup WireCommands.AppendSetup(type=APPEND_SETUP, requestId=5, segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, lastEventNumber=1442)
2019-01-24 18:22:56,023 693014 [epollEventLoopGroup-5-27] TRACE i.p.s.protocol.netty.CommandEncoder - Encoding message to send over the wire Append(segment=testReadWriteAndAutoScaleScope672520326/testReadWriteAndAutoScaleStream/7.#epoch.3, writerId=9a131fe4-2fd2-48f4-b69a-38f3aac98cb2, eventNumber=1443, eventCount=1, data=CompositeByteBuf(ridx: 0, widx: 55, cap: 55, components=2), expectedLength=null)
```

But after the last log line related to `CommandEncoder`, there is no more log activity related to writer `9a131fe4-2fd2-48f4-b69a-38f3aac98cb2`. So, the event was sent through the wire and written, but this was not noticed by the writer to get unblocked.

**Comment by RaulGracia** [comment]
Note that as far as I know this problem is impacting both `r0.4` and `r0.5` (see system test build 1491).

**Comment by shrids** [comment]
Checked the build 1491 logs. Observing that all the 4 writers spawned by the test `ReadTxnWriteAutoScaleWithFailoverTest`.

- Observing that an Append is sent out . But the ack (`WireCommands.DataAppended`) is never received by the client causing the writer to be blocked on flush().

- Log snippet for the last flush invocation by 1 of the writer after sending out the append is shown below.
```
2019-01-25 12:37:29,903 122445 [ReadTxnWriteAutoScaleWithFailoverTest-main-2] DEBUG i.p.t.system.AbstractReadWriteTest - Event write count before write call 1940
2019-01-25 12:37:29,903 122445 [ReadTxnWriteAutoScaleWithFailoverTest-main-2] TRACE i.p.c.s.i.SegmentOutputStreamImpl - Adding event 1097 to inflight on writer d7fe2eb7-6c89-4e39-91d7-243844966a05
2019-01-25 12:37:29,903 122445 [ReadTxnWriteAutoScaleWithFailoverTest-main-2] TRACE i.p.c.s.i.SegmentOutputStreamImpl - Sending append request: Append(segment=testReadTxnWriteAutoScaleScope1384583809/testReadTxnWriteAutoScaleStream/0.#epoch.0, writerId=d7fe2eb7-6c89-4e39-91d7-243844966a05, eventNumber=1097, eventCount=1, data=CompositeByteBuf(ridx: 0, widx: 54, cap: 54, components=2), expectedLength=null)
2019-01-25 12:37:29,903 122445 [epollEventLoopGroup-5-1] TRACE i.p.s.protocol.netty.CommandEncoder - Encoding message to send over the wire Append(segment=testReadTxnWriteAutoScaleScope1384583809/testReadTxnWriteAutoScaleStream/0.#epoch.0, writerId=d7fe2eb7-6c89-4e39-91d7-243844966a05, eventNumber=1097, eventCount=1, data=CompositeByteBuf(ridx: 0, widx: 54, cap: 54, components=2), expectedLength=null)
2019-01-25 12:37:29,903 122445 [ReadTxnWriteAutoScaleWithFailoverTest-main-2] DEBUG i.p.t.system.AbstractReadWriteTest - Event write count before flush 1940
2019-01-25 12:37:29,903 122445 [ReadTxnWriteAutoScaleWithFailoverTest-main-2] DEBUG i.p.c.s.i.SegmentOutputStreamImpl - Flushing writer: d7fe2eb7-6c89-4e39-91d7-243844966a05 with 1 inflight events
```
 - From the Segment store logs observing the following error 

```
2019-01-25 12:37:30,797 6644340 [core-18] ERROR i.p.s.server.writer.StorageWriter - StorageWriter[2]: Iteration[342].Error.
java.lang.IllegalArgumentException: handle must not be read-only.
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:122)
	at io.pravega.segmentstore.storage.rolling.RollingStorage.asWritableHandle(RollingStorage.java:785)
	at io.pravega.segmentstore.storage.rolling.RollingStorage.seal(RollingStorage.java:332)
	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$seal$3(AsyncStorageWrapper.java:98)
	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$runAsync$12(AsyncStorageWrapper.java:171)
	at io.pravega.segmentstore.storage.AsyncStorageWrapper.lambda$execute$13(AsyncStorageWrapper.java:184)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1590)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Complete logs: 
[readtxnwriteautoscalewithfailovertest-r5nse.zip](https://github.com/pravega/pravega/files/2802327/readtxnwriteautoscalewithfailovertest-r5nse.zip)
[segmentstore_2019-01-25.63.log.zip](https://github.com/pravega/pravega/files/2802375/segmentstore_2019-01-25.63.log.zip)

**Comment by tkaitchuck** [comment]
That last stack trace might be unrelated, it was occurring in the logs at regular intervals since well before the problematic message.

**Comment by tkaitchuck** [comment]
It looks like the following is going on:
On the server the append succeeded. This can be seen from the line: 
```
2019-01-25 12:37:29,908 6643451 [core-19] DEBUG i.p.s.s.logs.OperationProcessor - OperationProcessor[2]: CommitSuccess (TxnId = 1146, LastFullySerializedSN = 35029, LastStartedSN = 35029, Address = Sequence = 124554052730, LedgerId = 192, EntryId = 1146, Length = 166).
```
Following this we can see the `AutoScaleProcessor` getting invoked. This occurs inside of `AppendProcessor:313`. This means the ack callback `handleAppendResult` is getting invoked, without any exception. The call to `connection.send` just 5 lines down is never reached. Looking at the logs that thread is never heard from again. It must be blocked inside of the call `AutoScaleProcessor.writeRequest` which we can see gets invoked, but we never see the log of the return. 
Hence just as the test's client thread is blocked, it is actually blocked by a different thread internal to the server which is also blocked on performing a client write.
It is worth noting that we don't see any errors from that internal client, (or any logs at all). So it seems likely it is blocked on either the network send call (because the other side of the connection is not pulling data) or because it is blocked higher up in the stack on the write lock. In this case there must be some other thread holding it which is similarly blocked. 
I cannot pinpoint the root cause, but it likely has something to do with this recursive dependency interacting with the failure induced by the test somehow deadlocking the threads. 
If we could reproduce this and take a thread dump. (kill -3) it would be very helpful for debugging what is going on.

**Comment by fpj** [comment]
This issue is related to #3333, and the final solution will be addressed in that issue as per @tkaitchuck's comment.

**Comment by RaulGracia** [comment]
@fpj I think that this issue is independent of #3333. This issue has been fixed only in `r0.4` via PR #3335, but not yet in `master`.
On the other hand, issue #3333 is related to the internal logic used to mark readers in a reader group as offline, and its sporadic conflicts with Controller Event Processors under failover situations. This has been also fixed only in `r0.4` via PR #3336.
Note that at this point both issues may still appear in `master`. So, either we close this issue and create another one for `master` or we re-open this one to keep track of the problem.

**Comment by fpj** [comment]
@RaulGracia 

> I think that this issue is independent of #3333.

The change we merged in `0.4` is a workaround for the issue reported in #3333.

> This issue has been fixed only in `r0.4` via PR #3335, but not yet in `master`.

Yes, that's why the labels in this issue say `0.4.0`. The labels in #3333 refer to master (`0.5.0`) and an eventual bug-fix release of `0.4` (`0.4.1`).

> On the other hand, issue #3333 is related to the internal logic used to mark readers in a reader group as offline, and its sporadic conflicts with Controller Event Processors under failover situations. This has been also fixed only in `r0.4` via PR #3336.

See above.

> Note that at this point both issues may still appear in `master`. So, either we close this issue and create another one for `master` or we re-open this one to keep track of the problem.

Yes, which will be fixed with #3333.
